File: algorithms/3_hard/F/00154_find_minimum_in_rotated_sorted_array_ii/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static int findMin(const vector<int>& nums) {
        const isize len = std::ssize(nums);
        assert(len > 0);
        isize lo = 0;
        isize hi = len - 1;

        while (lo < hi) {
            const isize mid = lo + (hi - lo) / 2;
            if (nums[mid] < nums[hi]) {
                hi = mid;
            } else if (nums[mid] == nums[hi]) {
                hi -= 1;
            } else {
                lo = mid + 1;
            }
        }

        return nums[lo];
    }
};

File: algorithms/3_hard/F/00154_find_minimum_in_rotated_sorted_array_ii/solution.rs
impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        assert!(len > 0);
        let mut lo = 0_usize;
        let mut hi = len - 1;

        while lo < hi {
            let mid = lo + (hi - lo) / 2;
            if nums[mid] < nums[hi] {
                hi = mid;
            } else if nums[mid] == nums[hi] {
                hi -= 1;
            } else {
                lo = mid + 1;
            }
        }

        nums[lo]
    }
}
